### DISCUSSION TITLE
Confused by comment explaining puzzle/provider merge logic

### DISCUSSION THREAD

**Discussion by coffinmatician** [python-poetry/poetry#7049] [Q&A]
The following comment seems incorrect. https://github.com/python-poetry/poetry/blob/ba97fea031aecd5fd31878e15ece24b3ddc55752/src/poetry/puzzle/provider.py#L680-L688

My reading is that it should have `- foo (>=2.0 or <2.1) ; python_version >= "3.7"` instead of `- foo (>=2.0,<2.1) ; python_version >= "3.7"`

**Comment by neersighted** [comment]
Please use permalinks instead of branch/tag refs when linking source code, as otherwise it may move at any time, and future readers will have no idea what you were talking about. Also, please try to make your issues descriptive.

That being said, I don't really see the problem here. Are you confused by the syntax?

It's both valid in Poetry and follows PEP 508, as best I can tell:

```
>>> from poetry.core.constraints.generic import parse_constraint
>>> parse_constraint('foo (>=2.0,<2.1)')
<EmptyConstraint >
```

According to [PEP 508](https://peps.python.org/pep-0508/#grammar), the construct in question is a `version_many`.

```
version_cmp   = wsp* '<' | '<=' | '!=' | '==' | '>=' | '>' | '~=' | '==='
version       = wsp* ( letterOrDigit | '-' | '_' | '.' | '*' | '+' | '!' )+
version_one   = version_cmp version wsp*
version_many  = version_one (wsp* ',' version_one)*
versionspec   = ( '(' version_many ')' ) | version_many
urlspec       = '@' wsp* <URI_reference>
```

**Reply by coffinmatician** [reply]
https://peps.python.org/pep-0440/#version-specifiers seems to indicate that the `,` operator is an `and` operator. Therefore my reading is that the two versions I posted in my initial post aren't equivalent.

**Reply by neersighted** [reply]
Right, but the intersection here is `AND` as we are satisfying all the constraints, and are computing merged constraints whenever we can. `or` vs `,` doesn't really come into it in any meaningful way.

**Reply by coffinmatician** [reply]
So maybe I don't understand what this code is intending, but shouldn't `foo==2.3 and python_version==3.7` be a solution? Isn't that not a solution to `foo (>=2.0,<2.1) ; python_version >= "3.7"`, but it is a solution to:

```
                #   - foo (>=2.0) ; python_version >= "3.7"
                #   - foo (<2.1) ; python_version >= "3.7"
```

or does that no matter? I'm mostly asking looking into the solver as I have some projects that don't dependency resolve properly, but I am struggling to recreate the issue in a minified version. (The gist is that solving for "aarch64" with `markers` and secondary sources across multiple files (possibly being affected by `path` includes) isn't working.)

**Reply by neersighted** [reply]
I think that looking in the solver before creating a minimal reproducer is probably not a productive approach, if you don't already understand the solver.

Now, from the sound of things, you might be fundamentally misunderstanding how path dependencies work in Poetry. A path dependency does not pick up secondary sources defined in the Path dependencies pyproject.toml. This is because sources are a Poetry-made construct and they do not have a representation in the PEP 517 interface. When including a path dependency, Poetry treats it as a PEP 517 source tree, so that **any** Python project can serve as a path dependency. There is no special treatment of Poetry trees (this is actually a lie; we use Poetry to interpret it directly instead of invoking the `build-system`, but we don't leak any information we couldn't get by invoking Poetry as the `build-system`, so it's true enough in this case).

I'm still struggling to understand your concern around the comments -- are you asking if a version satisfies the constraint? It does not:

```
>>> from poetry.core.constraints.version.version import Version
>>> from poetry.core.constraints.version import parse_constraint
>>> version = Version.parse("2.3")
>>> parse_constraint(">=2.0").allows(version)
True
>>> parse_constraint("<2.0").allows(version)
False
>>> parse_constraint(">=2.0,<2.1").allows(version)
False
```

I think you're fundamentally misunderstanding what that code does -- it's trying to simplify constraints to reduce the number of items in the matrix before we start trying to find suitable versions that meet all our constraints.

In any case, it sounds like you're more likely dealing with a misunderstanding/misusage and not a bug in the solver; I'd suggest reaching out via Discord/Discussions so we can try and determine where you are going wrong and if it's a bug or not.

**Reply by coffinmatician** [reply]
I believe that I am trying to use Poetry correctly, but I will follow up later when I have a minimal example of it not working properly. (It is selecting a version that I have a `marker_machine == "x86_64"` on when on an aarch64 machine. It has one option from a `path`, and a second option from a secondary source. (The two options are a prebuilt wheel for x86_64, or a local path to build from source.))
